applications/vanilla/models/class.commentmodel.php

```diff
@@ -741,27 +741,25 @@ public function setWatch($discussion, $limit, $offset, $totalComments) {
 
         // If this discussion is in a category that has been marked read,
         // check if reading this thread causes it to be completely read again.
-        $categoryID = val('CategoryID', $discussion);
+        $categoryID = $discussion->CategoryID;
         if (!$categoryID) {
             return;
         }
         $category = CategoryModel::categories($categoryID);
         if (!$category) {
             return;
         }
-        $dateMarkedRead = val('DateMarkedRead', $category);
-        if (!$dateMarkedRead) {
-            return;
+        $wheres = ['CategoryID' => $categoryID];
+        $dateMarkedRead = $category->DateMarkedRead;
+        if ($dateMarkedRead) {
+            $wheres['DateLastComment>'] = $dateMarkedRead;
```

Comment:
I have some concerns about these changes. It is not 100% equal to previous approach.
Before if we had no `$dateMarkedRead` we returned early and we did not get discussions from `$discussionModel`.
Now we don't do early return and even more we are getting from the model with weaker filtering.
I think that can change user experience.

Reply:
I understand those concerns, but exiting at this point was the main reason why the functionality wasn't working as intended. This code blocks purpose is to determine if the category can be marked as read. "Categories that has never been marked as read weren't automatically marked as read when all discussions in such a discussion have been read."
This fix requires that `$dateMarkedRead` is handled in another way.

The UX was broken, therefore a change was needed.